src/compute/src/sink/persist_sink.rs

```diff
@@ -720,11 +731,38 @@ where
                                     persist_frontier
                                 );
                             }
-                            correction.with_correction_buffer(sink_metrics, sink_worker_metrics, |buffer| buffer.append(&mut data));
+
+                            if let Some(stash) = &mut desired_stash {
+                                for (d, t, r) in data {
+                                    stash.entry(t).or_insert_with(Vec::new).push((d, r));
+                                }
```

Comment:
I think you may want to consolidate `stash.entry(t)` under some circumstances. Otherwise retractions of future emissions will linger and consume resources until that time arrives. Standard pattern is to consolidate on each 2x growth we experience, which you can often do by sneaking a peek at `Vec::capacity()`: consolidate when "full" and post-consolidation double the capacity if it is not at least half-empty.

Reply:
I was wondering about that! I wasn't sure if we expect to see many future insertions/retractions that cancel each other out in Materialize. But I guess there is not much harm in adding this optimization, apart from some additional code complexity.